File: searchalgorithm.rb
#!/usr/bin/ruby -w

file = File.open("input.txt", 'r')

# hash for page and query details
pageHash={}
queryHash={}

# method for storing details from file to given hash of page and query 
def storeDetail(hashPage , hashQuery ,file)

indexP=1
indexQ=1

	while !file.eof?
      
		line = file.readline

		if (line[0]=="P")
			keyPage = "#{line[0]}#{indexP}"
			valuePage = readLine(line)
			indexP+=1
	  
			hashPage [keyPage] = valuePage
		end

		if (line[0]=="Q")
			keyQuery = "#{line[0]}#{indexQ}"
			valueQuery = readLine(line)
			indexQ+=1
	  
			hashQuery [keyQuery] = valueQuery
		end	
	
	end

end

#method to read Line and make key an there respective index hash
def readLine(line)

    max_Index = 8
    value=Hash.new{|k,v|}

    #removing first word from the line  
    mline=line.split(" ")[1..-1].join(" ")
      mline.split(" ").each do |word|
		
			k = word
			v = max_Index
            
			value [k] = v	    
		    max_Index-=1

	end

	return value

end

#invoking storeDetail method by supplying two hash and input file reference
storeDetail(pageHash , queryHash ,file)

#Storing Hash key List
queryKeysList = queryHash.keys
pageKeysList = pageHash.keys


#Comparing keywords of page and query stored in hash

#outer Query Hash 
@Qkey=0
for @Qkey in 0...queryKeysList.length

	resultMap = {}
	
	queryValueHash = queryHash.values
	
	qVKeys = queryValueHash[@Qkey].keys
	qValue = queryValueHash[@Qkey].values

	#Outer Page Hash
		@Pkey=0
	for @Pkey in 0...pageKeysList.length
	
		pageValueHash = pageHash.values
	
		pVKeys = pageValueHash[@Pkey].keys
		pValue = pageValueHash[@Pkey].values
		
		#Inner Query key hash
		indexStrength = 0
		@QGkey=0
		for @QGkey in 0...qVKeys.length
			
			#Inner Page Key Hash
			@PGkey=0
			for @PGkey in 0...pVKeys.length
			
				# Algorithm for Query Strength for Page
				localmulValue = 0
				
				if (qVKeys[@QGkey] ==pVKeys[@PGkey])
				
					@Qvalues = qValue[@QGkey]
					@Pvalues = pValue[@PGkey]
				
					localmulValue = @Qvalues * @Pvalues;
				
					indexStrength += localmulValue
				
					#update index for at least one page. (for multiple keys)				
					webPageKey = pageKeysList[@Pkey]
				
					resultMap[webPageKey] = indexStrength
				
				end
		
			end
			# Internal page loop end
		end
		# Internal query loop end
	
	end
	#Outer page loop end
	
	# Sorting Result Hash by Value an arrange in descending hash 
	finalHash = Hash[resultMap.sort_by {|k,v| v}.reverse]
		
	# Printing Sorted Page Results for respective Query 	
	print " #{queryKeysList[@Qkey]}: "
		
	@Jkey=0
	for @Jkey in 0...finalHash.length
	
	
	   print finalHash.keys[@Jkey]
	   print " "
	
	end
	
	puts " "
	
end
#Outer Query loop end

File: JavaProjects/src/com/prasad/hackerton/challange/SearchAlgorithm.java
//package com.mavenhive.test;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

/**
 * Page Search by Keywords using given search algorithm.
 */
 
public class SearchAlgorithm {

    public static int max_Keyword = 8; // N -- Max number of key word
    public static BufferedReader bin = null;
    public static String line = "";
    public static String splitBy = " ";

    public static int indexP = 1;
    public static int indexQ = 1;


    public static void main(String[] args) {

        // TreeMap for Storing PageDetails
        Map<String, TreeMap<String, Integer>> mapWebPage = new TreeMap<String, TreeMap<String, Integer>>();

        // TreeMap for Storing QueryDetails
        Map<String, TreeMap<String, Integer>> mapQuery = new TreeMap<String, TreeMap<String, Integer>>();

        // Taking Input From File Name: "inputdata.txt"
        //********************
        /*
        P Ford Car Review
        P Review Car
        P Review Ford
        P Toyota Car
        P Honda Car
        P Car
        Q Ford
        Q Car
        Q Review
        Q Ford Review
        Q Ford Car
        Q cooking French
        */
        //**********************

        storeDetails(mapWebPage, mapQuery);

        smartSearch(mapWebPage, mapQuery);

        System.out.println("");
        System.out.println("");

    }

    private static void storeDetails(Map<String, TreeMap<String, Integer>> mapWebPage, Map<String, TreeMap<String, Integer>> mapQuery) {
        try {
            bin = new BufferedReader(new FileReader("inputdata.txt"));
            line = bin.readLine(); // read only first line

            while (line != null) {
                String[] keywords = line.split(splitBy);

                // Inserting PageDetails
                if (keywords[0].equals("P")) {
                    mapWebPage.put("P" + indexP, readPage(line));

                    indexP++;
                }

                // Inserting QueryDetails
                if (keywords[0].equals("Q")) {
                    mapQuery.put("Q" + indexQ, readPage(line));

                    indexQ++;
                }

                line = bin.readLine(); // read after first line
            }


        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (bin != null) {
                try {
                    bin.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }


        }
    }

    // Method For Inserting Keywords Data in Internal TreeMap with Index
    private static TreeMap readPage(String line) {


        TreeMap<String, Integer> treeMap = new TreeMap<String, Integer>();

        String[] keywords = line.split(splitBy);

        if (keywords != null) {

            // Inserting Data in TreeMap for Given Key Details with Indexing to each Key
            for (int i = 1; i < keywords.length; i++) {
                treeMap.put(keywords[i], max_Keyword - i + 1);
            }

        }

//        System.out.println(" :" + treeMap);
        return treeMap;
    }


    private static void smartSearch(Map<String, TreeMap<String, Integer>> mapWebPage, Map<String, TreeMap<String, Integer>> mapQuery) {

        String Pkey = null;
        int Pvalues = 0;

        String webPageKey = null;

        //Query select
        for (Map.Entry<String, TreeMap<String, Integer>> mapList : mapQuery.entrySet()) {


            String queryKey = mapList.getKey();

            // Printing format for search Result
            System.out.println(" ");
            System.out.print(queryKey + ":");

            TreeMap<String, Integer> valuesQuery = mapList.getValue();

            // Map For Storing Page Details according to calculated strength
            TreeMap<String, Integer> resultMap = new TreeMap<String, Integer>();

            //page select
            for (Map.Entry<String, TreeMap<String, Integer>> mapPage : mapWebPage.entrySet()) {

                int indexStrength = 0;
                int passvalue = 0;

                webPageKey = mapPage.getKey();

                TreeMap<String, Integer> valuesPage = mapPage.getValue();

                // Internal query Details selected
                for (Map.Entry<String, Integer> list : valuesQuery.entrySet()) { //query key select
                    String Qkey = list.getKey();
                    Integer Qvalues = list.getValue();

                    //Internal Page Details selected
                    for (Map.Entry<String, Integer> listP : valuesPage.entrySet()) { //page key


                        Pkey = listP.getKey();
                        Pvalues = listP.getValue();

                        int localmulValue = 0;

                        // Logic for Comparing the key and allocating strength to that Page
                        if (Qkey.equals(Pkey)) {

                            localmulValue = Qvalues * Pvalues;

                            // to cover mutiple key in single page and need to select scope of variable carefully
                            indexStrength = indexStrength + localmulValue;

                            //update index for at least one page. (for multiple keys)
                            resultMap.put(webPageKey, indexStrength);

                            passvalue = indexStrength;

                        }

                    }//page internal loop end

                    if (Qkey.equals(Pkey)) {

                        // update index for same key but in different page
                        resultMap.put(webPageKey, passvalue);

                    }

                }// Query Internal loop end

            } //Page main loop end

            Map<Integer, String> map = sortMapByValues(resultMap);
            Set set = map.entrySet();
            Iterator iterator = set.iterator();

            while (iterator.hasNext()) {
                Map.Entry ansMap = (Map.Entry) iterator.next();
                System.out.print(ansMap.getKey() + " ");   // print sorted page in Output.
            }


        } //Query main loop end


    }

    private static HashMap sortMapByValues(TreeMap map) {
        List list = new LinkedList(map.entrySet());

        // Defining Custom Comparator
        Collections.sort(list, new Comparator() {
            public int compare(Object o1, Object o2) {

                // Reverse the comparison to print data in descending order of Index.
                return ((Comparable) ((Map.Entry) (o2)).getValue()).compareTo(((Map.Entry) (o1)).getValue());
            }
        });

        // Copying the sorted list in HashMap
        // LinkedHashMap is used to preserve the insertion order
        HashMap sortedHashMap = new LinkedHashMap();
        for (Iterator it = list.iterator(); it.hasNext(); ) {
            Map.Entry entry = (Map.Entry) it.next();
            sortedHashMap.put(entry.getKey(), entry.getValue());
        }
        return sortedHashMap;
    }


}
